### PR TITLE
refact(freetype): clean up global variables and infrequent functions

### PR DESCRIPTION
**Pull request by FASTSHIFT**
### Description of the feature or fix

1.Merge all global variables into `lv_freetype_context_t`.
2.Remove the support of cache size=0, because this performance is not good, it will hardly be used.
3.Clear the font pathname reference count multiplexing, this will not save much memory compared to freetype cache.
4.Canonical function naming `lv_ft_xxx` -> `lv_freetype_xxx`.
5.Remove `lv_ft_info_t` as input and output parameters, this is easy to make user error.
6.Optimize error log output, support viewing specific error causes.
7.Encapsulate `font_manager` on the basis of `lv_freetype`, including path management, `font-family` support, font resource reuse and recycling functions (implemented in the next PR).

### Checkpoints
- [x] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [x] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- [x] Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant. 
- [ ] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- [x] If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).
 
Be sure the following conventions are followed:
- [x] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [x] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [x] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [x] Use typed pointers instead of `void *` pointers
- [x] Do not `malloc` into a static or global variables. Instead declare the variable in `LV_ITERATE_ROOTS` list in [`lv_gc.h`](https://github.com/lvgl/lvgl/blob/master/src/misc/lv_gc.h) and mark the variable with `GC_ROOT(variable)` when it's used.
- [x] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [x] Widget members function must start with `lv_<modul_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [x] `struct`s should be used via an API and not modified directly via their elements.
- [x] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [x] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [x] Arguments must be named in H files too.
- [x] To register and use callbacks one of the followings needs to be followed: 
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

### CODE REVIEW

File: src/libs/freetype/lv_freetype.h [hunk 1]
```diff
@@ -23,52 +23,22 @@ extern "C" {
...
-/**
- * init freetype library
- * @param max_faces Maximum number of opened FT_Face objects managed by this cache instance. Use 0 for defaults.
- * @param max_sizes Maximum number of opened FT_Size objects managed by this cache instance. Use 0 for defaults.
- * @param max_bytes Maximum number of bytes to use for cached data nodes. Use 0 for defaults.
- *                  Note that this value does not account for managed FT_Face and FT_Size objects.
- * @return true on success, otherwise false.
- */
-bool lv_freetype_init(uint16_t max_faces, uint16_t max_sizes, uint32_t max_bytes);
+lv_res_t lv_freetype_init(uint16_t max_faces, uint16_t max_sizes, uint32_t max_bytes);
```
Comment (kisvegabor) [inline src/libs/freetype/lv_freetype.h:44]: Please do not remove the API comments.

**Review by kisvegabor — COMMENTED** [review]
@embeddedt What do you think about adding `freetype` to the CI? The only issue I see is it will be a extra dependencies and it might be complicated to install FreeType on Windows.

**Comment by embeddedt** [comment]
> it might be complicated to install FreeType on Windows

Do you mean for potentially running tests on Windows in the future, or for test case development?

**Comment by kisvegabor** [comment]
I mentioned Windows as some LVGL developers are using Windows and it's already difficult for them to the tests. Or can WSL2  manage `sudo apt-get install freetype` and we can say we support only Linux for testing?

**Comment by embeddedt** [comment]
> Or can WSL2 manage sudo apt-get install freetype and we can say we support only Linux for testing?

This is what I was thinking, yes. I find Linux distros' package management to be much simpler to work with than that of Windows. :laughing: 

Another option would be to detect the presence of FreeType in CMake and only enable the related tests if it is installed. I am unfamiliar with the current scripts used to collect all the test cases, so not sure how feasible that would be.

**Comment by kisvegabor** [comment]
Doh, it needs to be [installed manually](https://docs.lvgl.io/master/libs/freetype.html). Anyway, I hope it still works.

Could you update the CI in `master` to install FreeType? If so I added a basic FreeType test. After that we could rebase this PR on `master`.

File: src/libs/freetype/lv_freetype.h (hunk 1 above)
Comment (FASTSHIFT) [reply src/libs/freetype/lv_freetype.h:44]: done

**Comment by FASTSHIFT** [comment]
> Doh, it needs to be [installed manually](https://docs.lvgl.io/master/libs/freetype.html). Anyway, I hope it still works.
> 
> Could you update the CI in `master` to install FreeType? If so I added a basic FreeType test. After that we could rebase this PR on `master`.

We can directly compile the freetype source code with lvgl, which is the simplest and has no platform restrictions.
Just add in `CMakeLists.txt`:
```cmake
add_subdirectory(${PROJECT_SOURCE_DIR}/freetype2 ${PROJECT_BINARY_DIR}/build_freetype2)

target_link_libraries(
        Simulator PRIVATE
        lvgl
        lvgl::examples
        lvgl::demos
        lvgl::drivers
        freetype
)
```

**Comment by kisvegabor** [comment]
Hmmm, is's great! @embeddedt what do you think?

**Comment by embeddedt** [comment]
It sounds good to me as well. We just need to add some logic to fetch and extract a FreeType source tarball as otherwise we'd need to add it as a submodule, if I understand correctly.

**Comment by FASTSHIFT** [comment]
I tried to tailor the minimum configuration of freetype required by a lvgl, and adapted the file access and memory allocation of freetype to lvgl. 
Put the compressed file:
[freetype_port.zip](https://github.com/lvgl/lvgl/files/9951997/freetype_port.zip)
 in this directory:
image
and add it to `CMakeLists.txt`:
```cmake
add_definitions(
        -DFT2_BUILD_LIBRARY
        -DFT_CONFIG_MODULES_H=<freetype_port/ftmodule.h>
        -DFT_CONFIG_OPTIONS_H=<freetype_port/ftoption.h>
)
```

It takes up about 100KB of FLASH space on the MCU platform.

**Comment by kisvegabor** [comment]
Wao, I didn't know that FT can be configured to replace all memory and file system functions. What a great lib!

I wonder if we should promote this option in the docs too and add the content of `freetype_port.zip` into our `freetype` folder guarded with `LV_FREETYPE_USE_LVGL_PORT` or so.

What do you think?

**Comment by FASTSHIFT** [comment]
> Wao, I didn't know that FT can be configured to replace all memory and file system functions. What a great lib!
> 
> I wonder if we should promote this option in the docs too and add the content of `freetype_port.zip` into our `freetype` folder guarded with `LV_FREETYPE_USE_LVGL_PORT` or so.
> 
> What do you think?

This is a very good choice, I uploaded `lv_ftsystem.c`, but freetype does not call the functions in `lv_ftsystem.c` and uses its own default implementation. Can you help me check it?

**Comment by kisvegabor** [comment]
In the [CMakeList](https://github.com/freetype/freetype/blob/master/CMakeLists.txt) of FreeType I haven't found any mentions about `FT_CONFIG_OPTIONS_H` or `FT_CONFIG_OPTIONS_H` .

It seems it always wants to use its one `ftsystem.c`. See [here](https://github.com/freetype/freetype/blob/0f43a0e7ebfbda303f0b2e8b6d5db63c362233a3/CMakeLists.txt#L426-L432).

**Comment by FASTSHIFT** [comment]
> In the [CMakeList](https://github.com/freetype/freetype/blob/master/CMakeLists.txt) of FreeType I haven't found any mentions about `FT_CONFIG_OPTIONS_H` or `FT_CONFIG_OPTIONS_H` .

freetype will check the custom configuration file in [ftheader.h](https://github.com/freetype/freetype/blob/master/include/freetype/config/ftheader.h#L146), which is also mentioned in the [document](https://github.com/freetype/freetype/blob/master/docs/CUSTOMIZE#L114).

**Comment by FASTSHIFT** [comment]
> > In the [CMakeList](https://github.com/freetype/freetype/blob/master/CMakeLists.txt) of FreeType I haven't found any mentions about `FT_CONFIG_OPTIONS_H` or `FT_CONFIG_OPTIONS_H` .
> 
> freetype will check the custom configuration file in [ftheader.h](https://github.com/freetype/freetype/blob/master/include/freetype/config/ftheader.h#L146), which is also mentioned in the [document](https://github.com/freetype/freetype/blob/master/docs/CUSTOMIZE#L114).

We can recommend users to point the path to the lv_freetype directory:
```cmake
add_definitions(
        -DFT2_BUILD_LIBRARY
        -DFT_CONFIG_MODULES_H=<lvgl/src/libs/freetype/ftmodule.h>
        -DFT_CONFIG_OPTIONS_H=<lvgl/src/libs/freetype/ftoption.h>
)
```

**Comment by kisvegabor** [comment]
Great, thank you! So let's step one square back. What would be the bast way of build the freetype port in CI?

Probably `wget` freetype next to the lvgl folder in CI.  On desktop we can say freetype needs to be installed to run the tests.

**Comment by FASTSHIFT** [comment]
> Great, thank you! So let's step one square back. What would be the bast way of build the freetype port in CI?
> 
> Probably `wget` freetype next to the lvgl folder in CI. On desktop we can say freetype needs to be installed to run the tests.

I think you should try to use the latest version of the test, the installed version of freetype may be older versions not necessarily uniform.

**Comment by kisvegabor** [comment]
In this case we can have a `test_setup_enc.py` or script to install/download all we need. What do you think?

**Comment by FASTSHIFT** [comment]
> In this case we can have a `test_setup_enc.py` or script to install/download all we need. What do you think?

I agree with this.

**Comment by kisvegabor** [comment]
Okay, great. I'd like to merge it and I think we should open a new PR for updating the tests.

Just one thing: I'm not sure if it's really a good idea to use LVGL's FS function with FT. This way the example won't work if someone normally installed  FreeType. So I suggest making the use of LVGL FS optional with a defiine like `LV_FREETYPE_USE_LVGL_FS` and set it to `0` by default. What do you think?

**Comment by FASTSHIFT** [comment]
> Okay, great. I'd like to merge it and I think we should open a new PR for updating the tests.
> 
> Just one thing: I'm not sure if it's really a good idea to use LVGL's FS function with FT. This way the example won't work if someone normally installed FreeType. So I suggest making the use of LVGL FS optional with a defiine like `LV_FREETYPE_USE_LVGL_FS` and set it to `0` by default.

`LV_FREETYPE_USE_LVGL_PORT` is off by default, I can add notes in the document.

**Comment by kisvegabor** [comment]
Ah, I see now. Thank you.  and it's already documented [here](https://github.com/lvgl/lvgl/pull/3782/files#diff-89a31af9fc7085f81fc10827c928ca98e5e279b108792cabceba919b1dc6129dR60).

Ok, I merge it. Thank you @FASTSHIFT.

Are you interested in participating of updating the tests too?

**Comment by FASTSHIFT** [comment]
> Ah, I see now. Thank you. and it's already documented [here](https://github.com/lvgl/lvgl/pull/3782/files#diff-89a31af9fc7085f81fc10827c928ca98e5e279b108792cabceba919b1dc6129dR60).
> 
> Ok, I merge it. Thank you @FASTSHIFT.
> 
> Are you interested in participating of updating the tests too?

ok if I have free time I will do it ;)

**Comment by kisvegabor** [comment]
Great, thank you in advance!

**Pull request merged**